phpBB/phpbb/mention/source/user.php

```diff
@@ -0,0 +1,47 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license       GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\mention\source;
+
+class user extends base_user
+{
+	/**
+	 * {@inheritdoc}
+	 */
+	public function get_priority($row)
+	{
+		/*
+		 * Presence in array with all names for this type should not increase the priority
+		 * Otherwise names will not be properly sorted because we fetch them in batches
+		 * and the name from 'special' source can be absent from the array with all names
+		 * and therefore it will appear lower than needed
+		 */
+		return 0;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	protected function query($keyword, $topic_id)
+	{
+		$query = $this->db->sql_build_query('SELECT', [
+			'SELECT'    => 'u.username_clean, u.user_id',
+			'FROM'      => [
+				USERS_TABLE => 'u',
+			],
+			'WHERE'     => $this->db->sql_in_set('u.user_type', [USER_NORMAL, USER_FOUNDER]),
+			'ORDER_BY'  => 'u.user_lastvisit DESC'
```

Comment:
you can't select all users like that, you need to add some limit.
I don't think it is acceptable to load the entire user table and to cache it like that.
@phpbb/development-team what do you think

Reply:
Yeah I agree this will get way too big. What is this used for actually since `$keyword` is not even used in the query?

Reply:
@Derky I tried to find the right balance between the DB performance and the amount of cache files being saved. So that we do not have a cache file for every keyword/user/etc as well as we don't query the DB too many times during the topic display process.

Hmm, actually I stopped using the keyword in any query methods, the parameter could probably be removed then.
Do you have any suggestions for making it in a different way?

Reply:
well, you could add a where on the keyword and a limit statement. One thing to note here is that the more users we pull from the DB at once will reduce the number of requests to the API endpoint, so probably you want to balance that.

Reply:
@lavigor are you only using this to display the username in the post or also to create the list of options while typing?

If you only use it to display the username in the post, I advise to look at how we do that with quoting since we show the username as well. (I think it's just hardcoded in the post text)

Reply:
@Derky quite the opposite - this is only used to generate the list of options.

@CHItA how many cache files are supposed to be created? I can't imagine the exact amount for every typed letters and their combinations, but this will likely take more space then a single request to fetch them all.